python/src/main/java/org/apache/apex/malhar/python/base/ApexPythonEngine.java

```diff
@@ -0,0 +1,132 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.python.base;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+
+import org.apache.apex.malhar.python.base.requestresponse.PythonInterpreterRequest;
+import org.apache.apex.malhar.python.base.requestresponse.PythonRequestResponse;
+
+/**
+ * Defines the methods that needs to be implemented by the Python operator implementations. The first implementation
+ *  takes the approach of in-memory interpreter using JEP. Other possibilities are using Py4J which is an inter process
+ *   communication model.
+ */
+public interface ApexPythonEngine
+{
+  /***
+   * Used to perform any pre interpreter processing.
+   * @param preInitConfigs The configuration that is going to be used by the interpreter
+   * @throws ApexPythonInterpreterException if there is an issue in executing the pre interpreter logic
+   */
+  void preInitInterpreter(Map<PythonInterpreterConfig,Object> preInitConfigs) throws ApexPythonInterpreterException;
+
+  /***
+   * Starts the interpreter.
+   * @throws ApexPythonInterpreterException if library not locatable or any other issue starting the interpreter
+   */
+  void startInterpreter() throws ApexPythonInterpreterException;
+
+  /***
+   * Used to perform any logic that needs to be executed after the interpreter is started but before any tuples start
+   *  getting processed. Example, setting the starting state of the variables that are used in tuple processing.
+   * @throws ApexPythonInterpreterException
+   */
+  void postStartInterpreter() throws ApexPythonInterpreterException;
+
+  /***
+   * Runs a series of commands.
```

Comment:
Here the concept of a worker appears suddenly. How are such workers setup and managed? I assumed that an interpreter is either embedded or a single process?

Reply:
Added more javadocs to explain the concept of a worker. 

The workers are setup automatically by the python engine implementations. The number of workers is configurable by calling a setter that sets the number of threads in the worker thread pool.

Here is a brief details:

As part of the engine initialisation, a worker pool is created with each worker given one request queue and a response queue. The worker thread keeps listening on the request queues to perform work and submit the responses back. The Apex operator drains off the responses. The stragglers are emitted into a straggler port and the tuples which were processed within the SLAs are emitted into the normal port. The BasePythonExecution operator takes care of all of this logic. 

The javadocs inside InterpreterThread.java and InterpreterThreadWrapper.java gives a lot more details about the worker pool and related details. JepPythonEngine.java encapsulates a collection of InterterpreterWrapper instances ( which is a thin wrapper around a single worker a.k.a InterpreterThread).